Port to Qt 6

mitchcurtis opened the pull request:
The PatternSyntax enum had to be removed as QRegularExpression only
supports Perl-compatible regular expressions.

As QVariant's comparison operators were removed, a simplified
comparison is now done as a starting point, but should probably
be extended to support more types.

Fixes #84
Fixes #86

mitchcurtis committed:
Port to Qt 6

The PatternSyntax enum had to be removed as QRegularExpression only
supports Perl-compatible regular expressions.

As QVariant's comparison operators were removed, a simplified
comparison is now done as a starting point, but should probably
be extended to support more types.

Fixes #84
Fixes #86

mitchcurtis commented on tests/tst_rolesorter.qml:
```diff
@@ -60,12 +60,11 @@ Item {
-                var modelValue = testModel.get(i, sorter.roleName);
-                verify(modelValue === sorter.expectedValues[i],
-                       "Expected testModel value " + sorter.expectedValues[i] + ", actual: " + modelValue);
+            let actualValues = [];
```
This results in clearer error messages:

```
FAIL!  : SortFilterProxyModel::RoleSorterTests::test_roleSorters(mixedCaseStringRole) Compared values are not the same
   Actual   (): [A,C,D,b,e]
   Expected (): [A,b,C,D,e]
   Loc: [/Users/mitch/dev/SortFilterProxyModel/tests/tst_rolesorter.qml(67)]
```

mitchcurtis commented on tests/tst_stringsorter.qml:
```diff
@@ -75,12 +75,11 @@ Item {
+            for (var i = 0; i < testModel.count; i++) {
+                actualValues.push(testModel.get(i, sorter.roleName));
             }
+            compare(actualValues, sorter.expectedValues);
```
There is one failing test that I don't know how to fix:

```
FAIL!  : SortFilterProxyModel::StringSorterTests::test_stringSorters(doNotIgnorePunctuation) Compared values are not the same
   Actual   (): [a-a,aa,b-b,b-c,b.c,bc]
   Expected (): [aa,a-a,b.c,b-b,bc,b-c]
   Loc: [/Users/mitch/dev/SortFilterProxyModel/tests/tst_stringsorter.qml(82)]

```

oKcerG replied:
Yeah this test is finnicky, it depends on ICU. I think it passes on Windows

mitchcurtis replied:
I blacklisted it on macOS.

oKcerG commented:
About the QVariant compare, does it work for bools ? Could you add QDate/QTime/QDateTime support ?

mitchcurtis commented:
> About the QVariant compare, does it work for bools ? Could you add QDate/QTime/QDateTime support ?

Not currently, but I can make it. Not sure what it should do though... so I just made false < true.

mitchcurtis committed:
Update CMakeListst.txt for Qt 6

atraczyk committed:
cmake: fix c++17 build requirement for msc

mitchcurtis committed:
Merge pull request #1 from atraczyk/qt-6

cmake: fix c++17 build requirement for msvc

mitchcurtis committed:
Fix build with ASAN on Linux

I was getting the following error:

/usr/bin/ld: isle/3rdparty/SortFilterProxyModel/CMakeFiles/SortFilterProxyModel.dir/SortFilterProxyModel_autogen/mocs_compilation.cpp.o: relocation R_X86_64_PC32 against symbol `__asan_option_detect_stack_use_after_return' can not be used when making a shared object; recompile with -fPIC

And:

/home/mitch/dev/bgv/isle/3rdparty/SortFilterProxyModel/qqmlsortfilterproxymodel.cpp:1: error: In included file: "You must build your code with position independent code if Qt was configured with -reduce-relocations. "         "Compile your code with -fPIC (and not with -fPIE)."

Initially I fixed this with:

    set(CMAKE_POSITION_INDEPENDENT_CODE ON)

but was told the proper solution is to add the missing
target_link_libraries.

olafmandel commented:
Have an additional commit in [MenloSystems:qt6](https://github.com/MenloSystems/SortFilterProxyModel/tree/qt6) that allows compiling against Qt5 and Qt6 from the same codebase.

olafmandel commented:
> About the QVariant compare, does it work for bools ? Could you add QDate/QTime/QDateTime support ?

Which implementation is preferable: `qqsfpm::compareVariants()` from `utils/utils.*` or ` qqsfpm::lessThan()` from #81 ?

JoachimColine commented:
Why not use built-in variant comparison? https://doc.qt.io/qt-6/qvariant.html#compare

mitchcurtis commented:
Qt added support for SortFilterProxyModel in 6.10: https://doc-snapshots.qt.io/qt6-dev/qml-qtqml-models-sortfilterproxymodel.html
